### DISCUSSION TITLE
edit: solved

### DISCUSSION THREAD

**Discussion by BenWibking** [AMReX-Codes/amrex#2855] [Q&A]
Hi,

I have either discovered a compiler bug in gcc 10, 11, and 12, or there is very subtle undefined behavior in my code that is nonetheless works as expected when compiled by gcc 8.5, gcc 9.5, nvcc, and clang (with any optimization flags) or with gcc 10, 11, 12 when compiled at optimization levels -O0.  No asserts or warnings (except for unrecognized pragmas) are triggered with any compiler when building in debug mode.

Have any similar issues been seen with other AMReX codes? What is the best way to report this to the gcc developers (or rule out undefined behavior in my code)?

The issue is reproducible from this PR: https://github.com/BenWibking/quokka/pull/62

**Comment by BenWibking** [comment]
It was undefined behavior after all.

**Reply by WeiqunZhang** [reply]
I am curious.  Could you give us more detail on the UB?

**Reply by BenWibking** [reply]
The correct behavior depended on the constructor of an object being called before a subsequent assignment. The object was not used in between. In this case, gcc 10+ (but none of the other compilers, interestingly), optimizes out the call to the constructor. The code that caused it looked like this:

    // initialize all components to zero
    quokka::valarray<double, fluxdim> D_star(0.);
    D_star = {0., 1., 0., 0., S_star, 0., 0.};

The last line didn't fully initialize the object, so I got the wrong answer on a test problem.
